source/aws_signing.c

```diff
@@ -811,30 +890,51 @@ static int s_append_canonical_query_string(struct aws_uri *uri, struct aws_signi
 
     int result = AWS_OP_ERR;
 
+    /* aws-c-auth takes query params from the URI, so they should already be URI-encoded.
+     * To ensure that the signature uses "canonical" URI-encoding, we decode and then re-encode the params. */
+    struct aws_array_list temp_strings;
+    if (aws_array_list_init_dynamic(
+            &temp_strings, state->allocator, INITIAL_QUERY_FRAGMENT_COUNT * 4, sizeof(struct aws_string *))) {
```

Comment:
Feels like this could be more precisely calculated if done after the param split below